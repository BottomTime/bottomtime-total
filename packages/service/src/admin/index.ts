export * from './admin.service';

export * from './admin-tanks.controller';
export * from './admin-users.controller';

export * from './admin.module';
