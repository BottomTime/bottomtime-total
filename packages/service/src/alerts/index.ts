export * from './alert';
export * from './alerts.service';
export * from './alerts.module';
