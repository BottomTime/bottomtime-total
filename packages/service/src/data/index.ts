export * from './data-source';
export * from './get-query-vector';

export * from './certification.entity';
export * from './dive-site-review.entity';
export * from './dive-site.entity';
export * from './friend-request.entity';
export * from './friendship.entity';
export * from './tank.entity';
export * from './user-certification.entity';
export * from './user-oauth.entity';
export * from './user.entity';
