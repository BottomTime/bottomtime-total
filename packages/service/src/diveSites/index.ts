export * from './interfaces';

export * from './default-dive-site';
