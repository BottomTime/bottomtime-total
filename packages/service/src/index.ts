import { Config } from './config';
import { createApp } from './create-app';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then(async (app) => {
    await app.listen(Config.port);

    log.info(
      `ðŸŽ‰ Service has successfully started and is listening on port ${Config.port}. ðŸŽ‰`,
    );
  })
  .catch((error) => {
    log.fatal('[SERVICE] Failed to start service... Shutting down...', error);
    process.exit(1);
  });
