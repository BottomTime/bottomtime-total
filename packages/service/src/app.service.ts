import { AppMetricsDTO } from '@bottomtime/api';

import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';

import dayjs from 'dayjs';
import { MoreThan, Repository } from 'typeorm';

import { DiveSiteEntity, LogEntryEntity, UserEntity } from './data';

@Injectable()
export class AppService {
  constructor(
    @InjectRepository(UserEntity)
    private readonly users: Repository<UserEntity>,

    @InjectRepository(DiveSiteEntity)
    private readonly diveSites: Repository<DiveSiteEntity>,

    @InjectRepository(LogEntryEntity)
    private readonly logEntries: Repository<LogEntryEntity>,
  ) {}

  private async getUsersMetrics(): Promise<AppMetricsDTO['users']> {
    const [total, active, activeLastMonth] = await Promise.all([
      this.users.count(),
      this.users.countBy({
        lastLogin: MoreThan(dayjs().subtract(7, 'days').toDate()),
      }),
      this.users.countBy({
        lastLogin: MoreThan(dayjs().subtract(30, 'days').toDate()),
      }),
    ]);

    return {
      total,
      active,
      activeLastMonth,
    };
  }

  private async getSitesMetrics(): Promise<AppMetricsDTO['diveSites']> {
    const [total] = await Promise.all([this.diveSites.count()]);

    return {
      total,
    };
  }

  private async getLogEntriesMetrics(): Promise<AppMetricsDTO['logEntries']> {
    const [total, addedLastWeek, addedLastMonth] = await Promise.all([
      this.logEntries.count(),
      this.logEntries.countBy({
        createdAt: MoreThan(dayjs().subtract(14, 'days').toDate()),
      }),
      this.logEntries.countBy({
        createdAt: MoreThan(dayjs().subtract(30, 'days').toDate()),
      }),
    ]);

    return {
      total,
      addedLastWeek,
      addedLastMonth,
    };
  }

  async getMetrics(): Promise<AppMetricsDTO> {
    const [users, diveSites, logEntries] = await Promise.all([
      this.getUsersMetrics(),
      this.getSitesMetrics(),
      this.getLogEntriesMetrics(),
    ]);

    return {
      users,
      diveSites,
      logEntries,
    };
  }
}
