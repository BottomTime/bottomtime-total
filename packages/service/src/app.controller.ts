import { AppMetricsDTO, PingResponseDTO } from '@bottomtime/api';

import { Controller, Get, Inject } from '@nestjs/common';

import { AppService } from './app.service';

@Controller()
export class AppController {
  constructor(@Inject(AppService) private readonly service: AppService) {}

  @Get()
  getPing(): PingResponseDTO {
    return {
      api: 'Bottom Time Service',
      apiVersion: '1.0.0',
      uptime: Math.ceil(process.uptime() * 1000),
    };
  }

  @Get('api/metrics')
  async getMetrics(): Promise<AppMetricsDTO> {
    return await this.service.getMetrics();
  }
}
