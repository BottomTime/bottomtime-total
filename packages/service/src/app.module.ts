import { Module } from '@nestjs/common';
import { EventEmitterModule } from '@nestjs/event-emitter';
import { PassportModule } from '@nestjs/passport';
import { TypeOrmModule } from '@nestjs/typeorm';

import { DataSource, DataSourceOptions } from 'typeorm';

import { AdminModule } from './admin';
import { AlertsModule } from './alerts';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { AuthModule } from './auth/auth.module';
import { DiveSiteEntity, LogEntryEntity, UserEntity } from './data';
import { PostgresDataSourceOptions } from './data-source';
import { DiveSitesModule } from './diveSites/dive-sites.module';
import { EmailModule } from './email';
import { EventsModule } from './events';
import { FriendsModule } from './friends';
import { LogEntriesModule } from './logEntries';
import { MembershipModule } from './membership';
import { NotificationsModule } from './notifications';
import { OperatorsModule } from './operators';
import { TanksModule } from './tanks/tanks.module';
import { UsersModule } from './users';

@Module({
  imports: [
    EventEmitterModule.forRoot(),
    TypeOrmModule.forRootAsync({
      useFactory: () => PostgresDataSourceOptions,
      dataSourceFactory: async (options?: DataSourceOptions) => {
        const ds = new DataSource(options ?? PostgresDataSourceOptions);
        return await ds.initialize();
      },
    }),
    PassportModule.register({
      session: false,
    }),
    EventsModule,

    TypeOrmModule.forFeature([UserEntity, DiveSiteEntity, LogEntryEntity]),

    AlertsModule,
    AdminModule,
    AuthModule,
    OperatorsModule,
    DiveSitesModule,
    EmailModule,
    FriendsModule,
    LogEntriesModule,
    MembershipModule,
    NotificationsModule,
    TanksModule,
    UsersModule,
  ],

  providers: [AppService],
  controllers: [AppController],
})
export class AppModule {}
