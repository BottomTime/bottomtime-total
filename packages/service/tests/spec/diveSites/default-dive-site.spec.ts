import { Collection } from 'mongodb';

import { Collections, DiveSiteDocument } from '../../../src/data';
import { createTestLogger } from '../../test-logger';
import { DefaultDiveSite } from '../../../src/diveSites';
import { fakeDiveSite } from '../../fixtures/fake-dive-site';
import { mongoClient } from '../../mongo-client';

const log = createTestLogger('default-dive-site');

describe('Default Dive Site', () => {
  let Sites: Collection<DiveSiteDocument>;

  beforeAll(() => {
    Sites = mongoClient.db().collection(Collections.DiveSites);
  });

  it('Will return properties correctly', () => {
    const data = fakeDiveSite();
    const site = new DefaultDiveSite(mongoClient, log, data);
    expect(site.averageRating).toEqual(data.averageRating);
    expect(site.averageDifficulty).toEqual(data.averageDifficulty);
    expect(site.createdOn).toEqual(data.createdOn);
    expect(site.creator).toEqual(data.creator);
    expect(site.description).toEqual(data.description);
    expect(site.directions).toEqual(data.directions);
    expect(site.freeToDive).toEqual(data.freeToDive);
    expect(site.gps).toEqual({
      lat: data.gps!.coordinates[0],
      lon: data.gps!.coordinates[1],
    });
    expect(site.id).toEqual(data._id);
    expect(site.location).toEqual(data.location);
    expect(site.name).toEqual(data.name);
    expect(site.shoreAccess).toEqual(data.shoreAccess);
    expect(site.updatedOn).toEqual(data.updatedOn);
  });

  it('Will allow properties to be updated', () => {
    const oldData = fakeDiveSite();
    const newData = fakeDiveSite({
      freeToDive: !oldData.freeToDive,
      shoreAccess: !oldData.shoreAccess,
    });
    const site = new DefaultDiveSite(mongoClient, log, oldData);

    site.description = newData.description!;
    site.directions = newData.directions!;
    site.freeToDive = newData.freeToDive!;
    site.gps = {
      lat: newData.gps!.coordinates[0],
      lon: newData.gps!.coordinates[1],
    };
    site.location = newData.location;
    site.name = newData.name;
    site.shoreAccess = newData.shoreAccess!;

    expect(site.description).toEqual(newData.description);
    expect(site.directions).toEqual(newData.directions);
    expect(site.freeToDive).toEqual(newData.freeToDive);
    expect(site.gps).toEqual({
      lat: newData.gps!.coordinates[0],
      lon: newData.gps!.coordinates[1],
    });
    expect(site.location).toEqual(newData.location);
    expect(site.name).toEqual(newData.name);
    expect(site.shoreAccess).toEqual(newData.shoreAccess);
  });

  it('Will save a new dive site', async () => {
    const expected = fakeDiveSite();
    const diveSite = new DefaultDiveSite(mongoClient, log, expected);
    await diveSite.save();
    const actual = await Sites.findOne({ _id: expected._id });
    expect(actual).toEqual(expected);
  });

  it('Will update an existing dive site', async () => {
    const oldData = fakeDiveSite();
    const newData = fakeDiveSite({
      _id: oldData._id,
      creator: oldData.creator,
      createdOn: oldData.createdOn,
      updatedOn: oldData.updatedOn,
      averageRating: oldData.averageRating,
      reviews: oldData.reviews,
    });
    await Sites.insertOne(oldData);
    const site = new DefaultDiveSite(mongoClient, log, newData);

    await site.save();

    const actual = await Sites.findOne({ _id: newData._id });
    expect(actual).toEqual(newData);
  });

  it('Will remove fields that have been set to undefined when saving', async () => {
    const expected = fakeDiveSite();
    const site = new DefaultDiveSite(mongoClient, log, expected);
    await Sites.insertOne(expected);

    site.description = undefined;
    site.directions = undefined;
    site.freeToDive = undefined;
    site.gps = undefined;
    site.shoreAccess = undefined;

    delete expected.description;
    delete expected.directions;
    delete expected.freeToDive;
    delete expected.gps;
    delete expected.shoreAccess;

    await site.save();

    const actual = await Sites.findOne({ _id: expected._id });
    expect(actual).toEqual(expected);
  });

  it('Will delete a dive site', async () => {
    const data = fakeDiveSite();
    const site = new DefaultDiveSite(mongoClient, log, data);
    await Sites.insertOne(data);
    await site.delete();
    await expect(Sites.findOne({ _id: data._id })).resolves.toBeNull();
  });

  it('Will perform a no-op when deleting a site that does not exist', async () => {
    const data = fakeDiveSite();
    const site = new DefaultDiveSite(mongoClient, log, data);
    await expect(site.delete()).resolves.toBeUndefined();
  });

  it('Will export a JSON object', () => {
    const data: DiveSiteDocument = {
      _id: '3f83be2c-8d8a-4f06-a6b7-6bec1098ee11',
      creator: '35dc2551-2ebf-477a-995d-1a2e6f574e3a',
      createdOn: new Date('2018-08-19T08:56:21.007Z'),
      updatedOn: new Date('2022-12-08T23:06:27.525Z'),
      name: 'astonishing waste',
      description:
        'Dicta aliquam optio nesciunt error necessitatibus. Molestiae repellendus hic blanditiis voluptatibus.',
      location: 'Fayetteville Saint Martin',
      directions:
        'Et tempore impedit deleniti placeat ea asperiores in itaque.',
      gps: { type: 'Point', coordinates: [-0.05, -76.34] },
      freeToDive: false,
      shoreAccess: true,
      averageRating: 4.79,
      averageDifficulty: 2.28,
    };
    const site = new DefaultDiveSite(mongoClient, log, data);
    expect(site.toJSON()).toMatchSnapshot();
  });

  it('Will export a summarized JSON object', () => {
    const data: DiveSiteDocument = {
      _id: '363896cd-4310-416f-8861-47f3322af468',
      creator: 'a4fc342c-de93-4a6f-b433-8af6dbeadf90',
      createdOn: new Date('2022-04-11T13:04:56.199Z'),
      updatedOn: new Date('2022-06-09T21:00:40.101Z'),
      name: 'paltry linseed',
      description:
        'Modi beatae voluptates error. Magnam laboriosam quod dolores.',
      location: 'Coral Springs Hungary',
      directions: 'Dolorem at perspiciatis dolore maxime natus autem.',
      gps: { type: 'Point', coordinates: [45.19, 62.04] },
      freeToDive: true,
      shoreAccess: true,
      averageRating: 2.44,
      averageDifficulty: 1.81,
    };
    const site = new DefaultDiveSite(mongoClient, log, data);
    expect(site.toSummaryJSON()).toMatchSnapshot();
  });
});
