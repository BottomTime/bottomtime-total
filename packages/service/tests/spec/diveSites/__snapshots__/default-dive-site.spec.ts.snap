// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Dive Site Will export a JSON object 1`] = `
{
  "averageDifficulty": 2.28,
  "averageRating": 4.79,
  "createdOn": 2018-08-19T08:56:21.007Z,
  "creator": "Brad77",
  "description": "Dicta aliquam optio nesciunt error necessitatibus. Molestiae repellendus hic blanditiis voluptatibus.",
  "directions": "Et tempore impedit deleniti placeat ea asperiores in itaque.",
  "freeToDive": false,
  "gps": {
    "lat": -0.05,
    "lon": -76.34,
  },
  "id": "3f83be2c-8d8a-4f06-a6b7-6bec1098ee11",
  "location": "Fayetteville Saint Martin",
  "name": "astonishing waste",
  "shoreAccess": true,
  "updatedOn": 2022-12-08T23:06:27.525Z,
}
`;

exports[`Default Dive Site Will export a summarized JSON object 1`] = `
{
  "averageDifficulty": 1.81,
  "averageRating": 2.44,
  "createdOn": 2022-04-11T13:04:56.199Z,
  "creator": "Brad77",
  "freeToDive": true,
  "id": "363896cd-4310-416f-8861-47f3322af468",
  "location": "Coral Springs Hungary",
  "name": "paltry linseed",
  "shoreAccess": true,
  "updatedOn": 2022-06-09T21:00:40.101Z,
}
`;
